### PR TITLE
Breaking up transfers and draw calls

### PR DESCRIPTION
**Pull request by jherico**
I was able to get this mostly working.  However, for whatever reason, some of the draw draw calls dont work properly.  The ovlerays don't display (or display intermittently) and the overlays and skyboxes don't render properly.  I suspect it's a bad initial state.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: d9cc8781937b2e6b7a8493542516235a9b78796d available at: https://gist.github.com/15fa17a06da550790a63<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5539.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5539.dmg

File: libraries/gpu/src/gpu/GLBackend.cpp [hunk 1]
```diff
@@ -64,7 +66,7 @@ GLBackend::CommandCall GLBackend::_commandCalls[Batch::NUM_COMMANDS] =
 
 void GLBackend::init() {
     static std::once_flag once;
-    std::call_once(once, [] {
+    std::call_once(once, [&] {
```
Comment (ZappoMan) [inline libraries/gpu/src/gpu/GLBackend.cpp]: Why capture reference here?

File: libraries/gpu/src/gpu/GLBackend.cpp [hunk 2]
```diff
@@ -120,20 +122,98 @@ GLBackend::~GLBackend() {
...
+    _transform._objectTransforms.push_back(TransformObject());
+
+    _commandIndex = 0;
+    preUpdateTransform();
+
+    int drawCount = 0;
+    for (_commandIndex = 0; _commandIndex < numCommands; ++_commandIndex) {
+        switch (*command) {
+        case Batch::COMMAND_draw:
+        case Batch::COMMAND_drawIndexed:
```
Comment (ZappoMan) [inline libraries/gpu/src/gpu/GLBackend.cpp]: Coding standard: case inserted by 4 spaces from switch

**Comment by ZappoMan** [comment]
I assume the goal of this PR is to move toward reusing the batch for stereo. Right?

**Comment by jherico** [comment]
Yes, and reducing the time required to execute the GL calls.

File: libraries/gpu/src/gpu/GLBackend.h [hunk 3]
```diff
@@ -305,34 +310,36 @@ class GLBackend : public Backend {
...
     void updateTransform();
+    void preUpdateTransform();
     void resetTransformStage();
     struct TransformStageState {
         TransformObject _transformObject;
         TransformCamera _transformCamera;
-        GLuint _transformObjectBuffer;
-        GLuint _transformCameraBuffer;
+
+        QVector<TransformObject> _objectTransforms;
```
Comment (samcake) [inline libraries/gpu/src/gpu/GLBackend.h]: Can we stick to std::vector please

File: libraries/gpu/src/gpu/GLBackend.h [hunk 4]
```diff
@@ -305,34 +310,36 @@ class GLBackend : public Backend {
...
+        QVector<TransformCamera> _cameraTransforms;
```
Comment (samcake) [inline libraries/gpu/src/gpu/GLBackend.h]: std::vector please

File: libraries/gpu/src/gpu/GLBackendTransform.cpp [hunk 5]
```diff
@@ -98,33 +98,43 @@ void GLBackend::updateTransform() {
...
+            glBindBuffer(GL_UNIFORM_BUFFER, _transform._transformObjectBuffer);
+            glBindBufferRange(GL_UNIFORM_BUFFER, TRANSFORM_OBJECT_SLOT, _transform._transformObjectBuffer,
+                objOffset.second, sizeof(Backend::TransformObject));
+            _transform._objectOffsets.pop_front();
+        }
+    }
+    if (!_transform._cameraOffsets.empty()) {
+        const TransformStageState::Pair& camOffset = _transform._cameraOffsets.front();
+        if (_commandIndex == camOffset.first) {
+            glBindBuffer(GL_UNIFORM_BUFFER, _transform._transformCameraBuffer);
```
Comment (samcake) [inline libraries/gpu/src/gpu/GLBackendTransform.cpp]: I don;t think you need that bind here
THe next line glBindBUfferRange should be enough

File: libraries/gpu/src/gpu/GLBackendTransform.cpp [hunk 6]
```diff
@@ -98,33 +98,43 @@ void GLBackend::updateTransform() {
...
     // Flags are clean
     _transform._invalidView = _transform._invalidProj = _transform._invalidModel = _transform._invalidViewport = false;
 }
 
+void GLBackend::updateTransform() {
+    (void)CHECK_GL_ERROR();
+    if (!_transform._objectOffsets.empty()) {
+        const TransformStageState::Pair& objOffset = _transform._objectOffsets.front();
+        if (_commandIndex == objOffset.first) {
+            glBindBuffer(GL_UNIFORM_BUFFER, _transform._transformObjectBuffer);
```
Comment (samcake) [inline libraries/gpu/src/gpu/GLBackendTransform.cpp]: I don;t think you need that bind here
THe next line glBindBUfferRange should be enough

File: libraries/gpu/src/gpu/GLBackend.cpp (hunk 1 above)
Comment (jherico) [inline libraries/gpu/src/gpu/GLBackend.cpp]: originally I was going to init _uboAlignment here, before I realized that this method was static.  will remove.

File: libraries/gpu/src/gpu/GLBackend.cpp (hunk 2 above)
Comment (jherico) [inline libraries/gpu/src/gpu/GLBackend.cpp]: getting visual studio to do the right thing here is annoying.  will fix.

File: libraries/gpu/src/gpu/GLBackendTransform.cpp (hunk 6 above)
Comment (jherico) [inline libraries/gpu/src/gpu/GLBackendTransform.cpp]: fixed

File: libraries/gpu/src/gpu/GLBackendTransform.cpp (hunk 5 above)
Comment (jherico) [inline libraries/gpu/src/gpu/GLBackendTransform.cpp]: fixed

File: libraries/gpu/src/gpu/GLBackend.h (hunk 3 above)
Comment (jherico) [inline libraries/gpu/src/gpu/GLBackend.h]: fixed

File: libraries/gpu/src/gpu/GLBackend.h (hunk 4 above)
Comment (jherico) [inline libraries/gpu/src/gpu/GLBackend.h]: fixed

**Comment by hifi-gustavo** [comment]
Build console output for commit: 0c25063cd88d0a39c58bee3bc2d82e4b2ad50fdd available at: https://gist.github.com/15fa17a06da550790a63<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5539.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5539.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 8756d48747c0ac55655b3faf114dc5a1f8c75fb5 available at: https://gist.github.com/15fa17a06da550790a63<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5539.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5539.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 9c5cc16ba4e485f025c5e14288eeb017a6aab10a available at: https://gist.github.com/15fa17a06da550790a63<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5539.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5539.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: f4610152925c659c9d7e5b14abae94c3e9110fdb available at: https://gist.github.com/15fa17a06da550790a63<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5539.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5539.dmg

**Comment by jherico** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: f4610152925c659c9d7e5b14abae94c3e9110fdb available at: https://gist.github.com/15fa17a06da550790a63<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5539.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5539.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: f539e95e5345986024c07d7c310ff43a6e6f9b7a available at: https://gist.github.com/15fa17a06da550790a63<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5539.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5539.dmg

**Pull request merged**